gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,351 @@
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType(TableType.EXTERNAL_TABLE.toString());
+
+    if (partition.isPresent()) {
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    table.setSd(sd);
+    try {
+      this.client.createTable(table);
+      log.info("Created table " + tableName + " in db " + dbName);
+    } catch (TException e) {
+      log.error(String.format("Unable to create table %s from db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    }
+    return table;
+  }
+
+  public boolean existsTable(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      return this.client.tableExists(dbName, tableName);
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public boolean existsPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      client.getPartition(dbName, tableName, partitionValues);
+      return true;
+    } catch (NoSuchObjectException e) {
+      return false;
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of partition %s in table %s in db %s", partitionValues,
+          tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void dropTable(String dbName, String tableName) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      if (this.client.tableExists(dbName, tableName)) {
+        this.client.dropTable(dbName, tableName);
+        log.info("Dropped table " + tableName + " in db " + dbName);
+      }
+    } catch (TException e) {
+      log.error(String.format("Unable to deregister table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  public void dropPartition(String dbName, String tableName, List<String> partitionValues) {
+    Lock lock = locks.get(dbName + tableName);
+    lock.lock();
+
+    try {
+      client.dropPartition(dbName, tableName, partitionValues, false);
+      log.info("Dropped partition " + partitionValues + " in table " + tableName + " in db " + dbName);
+    } catch (NoSuchObjectException e) {
+      // Partition does not exist. Nothing to do
+    } catch (TException e) {
+      log.error(String.format("Unable to check existence of Hive partition %s in table %s in db %s", partitionValues,
+          tableName, dbName));
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private void addPartition(Table table, HivePartition partition, Path partitionLocation) throws TException {
+    Preconditions.checkArgument(table.getPartitionKeysSize() == partition.getValues().size());
```

Comment:
If someone creates a table that partitions on column A, and then try to add a partition whose partition key is A and B.

That being said I should probably compare the actual partition columns rather than the size.